KAFKA-3520: Add system tests for REST APIs of list connector plugins and config validation

Ishiihara opened the pull request:
@ewen @granders Ready for review.

Ishiihara committed:
Add system tests for REST APIs of list connector plugins and config validation

ewencp commented on tests/kafkatest/tests/connect/connect_plugins_rest_test.py:
```diff
@@ -0,0 +1,250 @@
+            'topics': {'name': 'topics', 'value': [], 'errors': [], 'recommended_values': [], 'visible': True}
+        }
+
+        configs = self.validate_config_def(self.FILE_SOURCE_CONNECTOR, self.FILE_SOURCE_ALL_CONFIGS, props)
```
This tests quite a bit which is largely redundant given the unit tests, right? System tests are more useful to make sure everything works when integrated together. This tests some of that (e.g. that you actually can validate against a real connector plugin), but a bunch of it seems to be validating details that could be confirmed via unit tests (e.g. validating in detail the output format). Would something much smaller and integrated with the existing `connect_rest_test.py` be enough here? That would also avoid adding a whole additional test (which has non-trivial setup/teardown overhead).

Ishiihara replied:
Well, we didn't actually verify against the real connector plugin in unit tests. This is partly because the test jars doesn't include any real connector plugins. and all we verified are against mock connector plugins. Also, the tests try to simulate how a user interacts with the configuration validation and try to ensure that in each step the expected number of errors and expected error messages are generated by the configuration validation.  Given that this test only takes roughly 2 minutes, the overhead seems reasonable. 

I think the gap here is that we don't have a test that ensures the connector can run given the correct configurations/cannot run given the in correct configurations.  Maybe we can add a test covering configuration validation as well as submitting the connector as ensure it runs? But ensuring the connector runs is already covered in other tests and configuration validation is covered in this test.

ewencp replied:
@Ishiihara Right, testing that it actually works with the plugins is good. I worry about testing a ton of details like matching every single field in the output -- that's better tested and iterated on via unit tests. If we evolve the format, e.g. adding new fields, it's more painful to validate that via system tests.

I'm not necessarily worried about adding one test (especially when it doesn't have an parameterized versions). I ask about this because system tests aren't like unit tests, so they should often encompass more functionality. Some of the stuff you're checking here, like error counts, can be checked in unit tests of AbstractHerder (where this count is computed).

You're right that checking actual instantiation after validating (or failing validation of) the config would also be good. This would actually fit quite nicely with the existing REST test, which also just exercises basic listing, instantiation, configs, and deletion. That test has a similar issue in that it is checking exact equivalence of expected output from the API; these checks only worry me a bit more because there's so much more to them. Might make sense to live with those more extensive checks for now, at the cost of some fragility, but just refactor some of this into the existing REST test.

ewencp commented on tests/kafkatest/tests/connect/connect_plugins_rest_test.py:
```diff
@@ -0,0 +1,250 @@
+        self.validate_config_value(expected_config_values, actual_config_values)
+
+    def validate_plugins(self, expected, actual):
+        assert expected == actual
```
No need for this method since it's just an assert. I had to scroll all the way down here just to figure out if it did something different...

Ishiihara committed:
Add configuration validation to REST API test

Ishiihara committed:
Remove connect_plugins_rest_test

Ishiihara commented:
This depends on https://github.com/apache/kafka/pull/1356

Ishiihara commented:
@ewencp Ready for another round of review. Thanks!

ewencp commented on tests/kafkatest/services/connect.py:
```diff
@@ -140,6 +141,12 @@ def pause_connector(self, name, node=None):
         return self._rest('/connectors/' + name + '/resume', method="PUT")
 
+    def list_connector_plugins(self, node=None, retries=0, retry_backoff=.01):
+        return self._rest_with_retry('/connector-plugins/', node=node, retries=retries, retry_backoff=retry_backoff)
```
This should never hit a 409, so `_rest_with_retry` shouldn't be necessary as far as I can tell.

ewencp commented on tests/kafkatest/services/connect.py:
```diff
@@ -140,6 +141,12 @@ def pause_connector(self, name, node=None):
+        return self._rest_with_retry('/connector-plugins/', node=node, retries=retries, retry_backoff=retry_backoff)
+
+    def validate_config(self, connector_type, validate_request, node=None, retries=0, retry_backoff=.01):
+        return self._rest_with_retry('/connector-plugins/' + connector_type + '/config/validate', validate_request, node=node, method="PUT", retries=retries, retry_backoff=retry_backoff)
```
Same here re: `_rest_with_retry` since validation just uses a temporary connector instance to test the config.

ewencp commented on tests/kafkatest/tests/connect/connect_rest_test.py:
```diff
@@ -15,18 +15,47 @@
+
+    COMMON_GROUP = 'Common'
+    # The value denote whether the config is required or not.
+    COMMON_CONNECTOR_CONFIGS = {
```
I think this came up during unit test review too -- this is pretty exhaustive and detailed for a system test. Rather than evaluating high level functionality, this seems like it is evaluating all the minute details. Do we need this level of detail in the system test and is checking, e.g., the specific value of `importance` (rather than, for example, just that the field exists) going to add much to this test?

At a high level, I think we want this test to validate the API fails or succeeds with some basic sets of values. We've gotten away with `==` checks on some things so far because the amount of info was pretty minimal or trivially defined. but I don't know that we want things quite as exacting as they are here -- any minor change to the API, set of fields included, etc will cause this to break. This makes the system tests a lot harder to maintain. (And in fact I think some of the other checks should probably be made a bit more liberal).

I know we need to construct a valid set of configs for each connector, but maybe we should loosen up the assertions here? Maybe split it into fields we _require_ to have certain values (which is probably just one or two key ones) and fields we expect to exist (but don't assert specific values for)? Maybe we can find a balance that's more reasonable to maintain given the possibility of extensions to the current set of ConfigDef properties but still validates the API works as expected?

ewencp commented on tests/kafkatest/services/connect.py:
```diff
@@ -52,7 +52,7 @@ class ConnectServiceBase(KafkaPathResolverMixin, Service):
     }
 
-    def __init__(self, context, num_nodes, kafka, files):
+    def __init__(self, context, num_nodes, kafka, files=[]):
```
Are we adding these defaults for a reason? I don't see how these are being taken advantage of and in general, object defaults are dangerous since a single instance is shared across all calls (so, e.g., if a list is mutated, all subsequent calls will see the modified value).

ewencp commented:
@Ishiihara Also, I kicked off a test run here https://jenkins.confluent.io/job/system-test-kafka-branch-builder/441/console but some of the tests are failing.

Ishiihara committed:
Address review comments

Ishiihara commented:
@ewencp Addressed the review comments. The test failures seems not related to this change.  Also, started another test run https://jenkins.confluent.io/job/system-test-kafka-branch-builder/444/console

ewencp commented:
@Ishiihara I don't immediately see how they are related, but I also don't see the same failures on tests from other Jenkins builds and I am seeing the same errors in the latest test build you kicked off. Are you seeing these same test errors somewhere else that would suggest they are not an artifact of the changes in this patch?

I dug up the logs from one of the test runs on Jenkins and see some logs like this:

```
[2016-05-10 06:01:05,648] ERROR Failed to create job for /mnt/connect/connect-connector-0.properties (org.apache.kafka.connect.cli.ConnectStandalone)
[2016-05-10 06:01:05,649] ERROR Stopping after connector error (org.apache.kafka.connect.cli.ConnectStandalone)
java.util.concurrent.ExecutionException: org.apache.kafka.connect.errors.ConnectException: Failed to find any class that implements Connector and which name matches  available connectors are: org.apache.kafka.connect.sink.SinkConnector, org.apache.kafka.connect.file.FileStreamSourceConnector, org.apache.kafka.connect.tools.VerifiableSinkConnector, org.apache.kafka.connect.tools.VerifiableSourceConnector, org.apache.kafka.connect.source.SourceConnector, org.apache.kafka.connect.file.FileStreamSinkConnector
        at org.apache.kafka.connect.util.ConvertingFutureCallback.result(ConvertingFutureCallback.java:80)
        at org.apache.kafka.connect.util.ConvertingFutureCallback.get(ConvertingFutureCallback.java:67)
        at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:97)
Caused by: org.apache.kafka.connect.errors.ConnectException: Failed to find any class that implements Connector and which name matches  available connectors are: org.apache.kafka.connect.sink.SinkConnector, org.apache.kafka.connect.file.FileStreamSourceConnector, org.apache.kafka.connect.tools.VerifiableSinkConnector, org.apache.kafka.connect.tools.VerifiableSourceConnector, org.apache.kafka.connect.source.SourceConnector, org.apache.kafka.connect.file.FileStreamSinkConnector
        at org.apache.kafka.connect.runtime.Worker.getConnectorClass(Worker.java:226)
        at org.apache.kafka.connect.runtime.Worker.startConnector(Worker.java:166)
        at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.startConnector(StandaloneHerder.java:250)
        at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.putConnectorConfig(StandaloneHerder.java:164)
        at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:94)
[2016-05-10 06:01:05,650] INFO Kafka Connect stopping (org.apache.kafka.connect.runtime.Connect)
[2016-05-10 06:01:05,650] INFO Stopping REST server (org.apache.kafka.connect.runtime.rest.RestServer)
[2016-05-10 06:01:05,653] INFO Stopped ServerConnector@639bb83d{HTTP/1.1}{0.0.0.0:8083} (org.eclipse.jetty.server.ServerConnector)
[2016-05-10 06:01:05,659] INFO Stopped o.e.j.s.ServletContextHandler@401c88af{/,null,UNAVAILABLE} (org.eclipse.jetty.server.handler.ContextHandler)
[2016-05-10 06:01:05,660] INFO REST server stopped (org.apache.kafka.connect.runtime.rest.RestServer)
[2016-05-10 06:01:05,660] INFO Herder stopping (org.apache.kafka.connect.runtime.standalone.StandaloneHerder)
[2016-05-10 06:01:05,660] INFO Stopping connector local-file-source (org.apache.kafka.connect.runtime.Worker)
[2016-05-10 06:01:05,660] ERROR Error shutting down connector local-file-source:  (org.apache.kafka.connect.runtime.standalone.StandaloneHerder)
org.apache.kafka.connect.errors.ConnectException: Connector local-file-source not found in this worker.
        at org.apache.kafka.connect.runtime.Worker.stopConnector(Worker.java:282)
        at org.apache.kafka.connect.runtime.standalone.StandaloneHerder.stop(StandaloneHerder.java:85)
        at org.apache.kafka.connect.runtime.Connect.stop(Connect.java:71)
        at org.apache.kafka.connect.cli.ConnectStandalone.main(ConnectStandalone.java:101)
[2016-05-10 06:01:05,660] INFO Worker stopping (org.apache.kafka.connect.runtime.Worker)
[2016-05-10 06:01:05,660] WARN Shutting down tasks [] uncleanly; herder should have shut down tasks before the Worker is stopped. (org.apache.kafka.connect.runtime.Worker)
[2016-05-10 06:01:05,660] INFO Stopped FileOffsetBackingStore (org.apache.kafka.connect.storage.FileOffsetBackingStore)
[2016-05-10 06:01:05,660] INFO Worker stopped (org.apache.kafka.connect.runtime.Worker)
[2016-05-10 06:01:05,660] INFO Herder stopped (org.apache.kafka.connect.runtime.standalone.StandaloneHerder)
[2016-05-10 06:01:05,660] INFO Kafka Connect stopped (org.apache.kafka.connect.runtime.Connect)
```

The empty string for the connector class seems to indicate there might be some issue getting the info into the template with the `FILE_SINK_CONNECTOR` reference. I don't understand the exact problem yet, but I think that might be the source of problems w/ tests on this branch.

Ishiihara committed:
Address system test failure

Ishiihara commented:
@ewencp  Sorry, I think `FILE_SOURCE_CONNECTOR` and `FILE_SINK_CONNECTOR` are not defined in `connect_test.py` and `connect_distributed_test.py`. Made the necessary change and test passed locally. Also kicked another test run
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/445/console

ewencp commented:
@Ishiihara Since one of the tests is hanging, do you want to re-run this with just the connect/ folder of tests?

Ishiihara commented:
@ewencp Started another one with connect tests https://jenkins.confluent.io/job/system-test-kafka-branch-builder/447/console

Ishiihara commented:
@ewencp Test passed https://jenkins.confluent.io/job/system-test-kafka-branch-builder/447/console

ewencp commented:
LGTM, thanks @Ishiihara!
